docling/document_extractor.py

```diff
@@ -0,0 +1,331 @@
+import hashlib
+import logging
+import sys
+import threading
+import time
+from collections.abc import Iterable, Iterator
+from concurrent.futures import ThreadPoolExecutor
+from functools import partial
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple, Type, Union
+
+from pydantic import BaseModel, ConfigDict, model_validator, validate_call
+
+from docling.backend.abstract_backend import AbstractDocumentBackend
+from docling.backend.asciidoc_backend import AsciiDocBackend
+from docling.backend.csv_backend import CsvDocumentBackend
+from docling.backend.docling_parse_v4_backend import DoclingParseV4DocumentBackend
+from docling.backend.html_backend import HTMLDocumentBackend
+from docling.backend.json.docling_json_backend import DoclingJSONBackend
+from docling.backend.md_backend import MarkdownDocumentBackend
+from docling.backend.mets_gbs_backend import MetsGbsDocumentBackend
+from docling.backend.msexcel_backend import MsExcelDocumentBackend
+from docling.backend.mspowerpoint_backend import MsPowerpointDocumentBackend
+from docling.backend.msword_backend import MsWordDocumentBackend
+from docling.backend.noop_backend import NoOpBackend
+from docling.backend.pypdfium2_backend import PyPdfiumDocumentBackend
+from docling.backend.xml.jats_backend import JatsDocumentBackend
+from docling.backend.xml.uspto_backend import PatentUsptoDocumentBackend
```

Comment:
at the moment this will work only on PDF and images, right?
why all of these backend imports?
if we add a new backend, is this a list to be maintained as well?

Reply:
@dolfim-ibm I think we will need a "pure text" pipeline for extraction for the programmatic use-cases. I think we can "abuse" the current VLM by not providing it an image and serializing the converted document into markdown, but this looks suboptimal.

Reply:
@PeterStaar-IBM I would keep "text input to extraction model" out of this PR. There are opportunities to extend this for example by:
- Adding ExtractionPipelineOptions to control if the input that goes to the extraction model should be page images or the converted document output.
- Adding other convenience `extract_...` APIs that may accept directly a string and not an `InputDocument` type.